### ISSUE TITLE
Trivial/degenerate solution to Wav2vec 2

### ISSUE THREAD

**Issue by dxli94** [pytorch/fairseq#3735]
### Before asking:   
1. search the issues.   
2. search the docs.    

#### What is your question?

I was trying to understand Wav2Vec 2.0, and it seems the implementation might lead to trivial solutions in some cases. 

Specifically, if the model always assigns positive and negative samples with the same code, it might get very good InfoNCE estimation.  

The reason I believe lies in the implementation of compute_pred() in https://github.com/pytorch/fairseq/blob/master/fairseq/models/wav2vec/wav2vec2.py#L478, where if neg_is_pos() evaluates True, the negative will be assigned with -inf value as for logits. In this case, the cross-entropy loss in wav2vec_criterion.py will be trivially minimized, despite that the learnt representation may not be necessarily meaningful.

I wonder is this something by design? Did you encounter this issue before and have any experience on how to avoid it?

Your help is very much appreciated.

#### Code

   

```python
    def compute_preds(self, x, y, negatives):

        neg_is_pos = (y == negatives).all(-1)
        y = y.unsqueeze(0)
        targets = torch.cat([y, negatives], dim=0)

        logits = torch.cosine_similarity(x.float(), targets.float(), dim=-1).type_as(x)

        logits = logits / self.logit_temp

        if is_xla_tensor(logits) or neg_is_pos.any():
            fillval = -float(2 ** 30)
            if not hasattr(self, "_inftensor"):
                self._inftensor = (
                    torch.tensor(fillval).to(x.device)
                    if is_xla_tensor(logits)
                    else float("-inf")
                )
            logits[1:] = index_put(logits[1:], neg_is_pos, self._inftensor)

        return logits
```

#### What's your environment?

 - fairseq Version (master):
 - PyTorch Version (1.7)
 - OS (e.g., Linux): ubuntu 18
 - How you installed fairseq (`pip`, source): source
 - Build command you used (if compiling from source): pip install --editable ./
 - Python version: 3.7
 - CUDA/cuDNN version: 10.1
 - GPU models and configuration: 2080 Ti
 - Any other relevant information:

**Comment by KinWaiCheuk** [comment]
I am facing the same problem. I am training wav2vec with my own dataset using the default `wav2vec2_base.yaml` configuration, but after some time, the training accuracy would drops to zero while the validation accuracy increase to 1.
<img width="335" alt="image" src="https://user-images.githubusercontent.com/37035802/128176698-6b8a1847-5ecb-4b81-8114-b2a7525e680b.png">


The code perplexity and loss_0 do not look right either
<img width="326" alt="image" src="https://user-images.githubusercontent.com/37035802/128176867-f85b1ce4-c08c-4b56-bf80-d6f213609a3b.png"> <img width="324" alt="image" src="https://user-images.githubusercontent.com/37035802/128176889-1085f435-4705-43aa-b3c5-071471ce514e.png">


When I try to load the best checkpoint and see what feature wav2vec extracted, I see the output full of `-inf`


```python
model(x)['x']
output:
tensor([[[9.9857, 9.9857, 9.9857,  ..., 9.9857, 9.9857, 9.9857],
          [9.9857, 9.9857, 9.9857,  ..., 9.9857, 9.9857, 9.9857],
          [9.9857, 9.9857, 9.9857,  ..., 9.9857, 9.9857, 9.9857],
          [9.9857, 9.9857, 9.9857,  ..., 9.9857, 9.9857, 9.9857]],
 
         [[  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf],
          [  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf],
          [  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf],
          [  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf]],
 
         [[  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf],
          [  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf],
          [  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf],
          [  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf]],
 
         ...,
 
         [[  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf],
          [  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf],
          [  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf],
          [  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf]],
 
         [[  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf],
          [  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf],
          [  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf],
          [  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf]],
 
         [[  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf],
          [  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf],
          [  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf],
          [  -inf,   -inf,   -inf,  ...,   -inf,   -inf,   -inf]]],
        grad_fn=<CopySlices>
```

When I try to get the validation feature, the features are all the same across all time steps
```python
model(x, mask=False, features_only=True)['x']
output:
tensor([[[ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         ...,
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728]],

        [[ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         ...,
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728]],

        [[ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         ...,
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728]],

        [[ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         ...,
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728],
         [ 0.5195,  1.2744,  1.0260,  ..., -0.6923,  0.0550, -0.5728]]],
       grad_fn=<TransposeBackward0>)
```

My environment is?

fairseq Version (master):
PyTorch Version (1.9)
OS (e.g., Linux): ubuntu 18
How you installed fairseq (pip, source): source
Build command you used (if compiling from source): pip install --editable ./
Python version: 3.8.10
CUDA/cuDNN version: 11.2
GPU models and configuration: Tesla V100
Any other relevant information:
I am training wav2vec with my own dataset, where the average audio lengths are around 7 minutes.

**Comment by lixx0105** [comment]
@dxli94 @KinWaiCheuk hi，do you have any solutions for this problem? thanks.

**Comment by KinWaiCheuk** [comment]
> @dxli94 @KinWaiCheuk hi，do you have any solutions for this problem? thanks.

Currently, I reduce the learning rate from 0.0005 to 0.0001 to avoid the degenerate solution.
